### PR TITLE
Use assertTrueEventually in AuthenticationInformationLeakTest API-1652

### LINKED ISSUE #22564: com.hazelcast.client.AuthenticationInformationLeakTest.testAuthenticationExceptionDoesNotLeakInfo [API-1652]
_master_ (commit a2ab6c6ab6e09865dff34e035d301c095d875ad1)
Failed on IbmJDK8-fips: https://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-master-IbmJDK8-fips/26/testReport/junit/com.hazelcast.client/AuthenticationInformationLeakTest/testAuthenticationExceptionDoesNotLeakInfo/
<details><summary>Stacktrace:</summary>

```
java.io.EOFException
	at com.hazelcast.internal.nio.IOUtil.readFully(IOUtil.java:285)
	at com.hazelcast.client.impl.clientside.ClientTestUtil.readResponse(ClientTestUtil.java:78)
	at com.hazelcast.client.AuthenticationInformationLeakTest.getKeyValue(AuthenticationInformationLeakTest.java:170)
	at com.hazelcast.client.AuthenticationInformationLeakTest.testAuthenticationExceptionDoesNotLeakInfo(AuthenticationInformationLeakTest.java:99)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:90)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
	at java.lang.reflect.Method.invoke(Method.java:508)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:115)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:107)
	at java.util.concurrent.FutureTask.run(FutureTask.java:277)
	at java.lang.Thread.run(Thread.java:825)

```
</details>
<details><summary>Standard output:</summary>

```
Finished Running Test: testFailedAuthenticationDoesNotLeakInfoSerializationVersionMismatch in 2.051 seconds.
Started Running Test: testAuthenticationExceptionDoesNotLeakInfo
12:08:12,734  INFO |testAuthenticationExceptionDoesNotLeakInfo| - [MetricsConfigHelper] main - [LOCAL] [45c8778d-641e-49aa-a0f7-5e075f9348dd] [5.3.0-SNAPSHOT] Overridden metrics configuration with system property 'hazelcast.metrics.collection.frequency'='1' -> 'MetricsConfig.collectionFrequencySeconds'='1'
12:08:12,735  INFO |testAuthenticationExceptionDoesNotLeakInfo| - [logo] main - [127.0.0.1]:5701 [45c8778d-641e-49aa-a0f7-5e075f9348dd] [5.3.0-SNAPSHOT] 
	+       +  o    o     o     o---o o----o o      o---o     o     o----o o--o--o
	+ +   + +  |    |    / \       /  |      |     /         / \    |         |   
	+ + + + +  o----o   o   o     o   o----o |    o         o   o   o----o    |   
	+ +   + +  |    |  /     \   /    |      |     \       /     \       |    |   
	+       +  o    o o       o o---o o----o o----o o---o o       o o----o    o   
12:08:12,735  INFO |testAuthenticationExceptionDoesNotLeakInfo| - [system] main - [127.0.0.1]:5701 [45c8778d-641e-49aa-a0f7-5e075f9348dd] [5.3.0-SNAPSHOT] Copyright (c) 2008-2022, Hazelcast, Inc. All Rights Reserved.
12:08:12,735  INFO |testAuthenticationExceptionDoesNotLeakInfo| - [system] main - [127.0.0.1]:5701 [45c8778d-641e-49aa-a0f7-5e075f9348dd] [5.3.0-SNAPSHOT] Hazelcast Platform 5.3.0-SNAPSHOT (20221020 - a2ab6c6) starting at [127.0.0.1]:5701
12:08:12,735  INFO |testAuthenticationExceptionDoesNotLeakInfo| - [system] main - [127.0.0.1]:5701 [45c8778d-641e-49aa-a0f7-5e075f9348dd] [5.3.0-SNAPSHOT] Cluster name: 45c8778d-641e-49aa-a0f7-5e075f9348dd
12:08:12,735  INFO |testAuthenticationExceptionDoesNotLeakInfo| - [system] main - [127.0.0.1]:5701 [45c8778d-641e-49aa-a0f7-5e075f9348dd] [5.3.0-SNAPSHOT] Integrity Checker is disabled. Fail-fast on corrupted executables will not be performed. For more information, see the documentation for Integrity Checker.
12:08:12,735  INFO |testAuthenticationExceptionDoesNotLeakInfo| - [system] main - [127.0.0.1]:5701 [45c8778d-641e-49aa-a0f7-5e075f9348dd] [5.3.0-SNAPSHOT] The Jet engine is disabled.
To enable the Jet engine on the members, do one of the following:
  - Change member config using Java API: config.getJetConfig().setEnabled(true)
  - Change XML/YAML configuration property: Set hazelcast.jet.enabled to true
  - Add system property: -Dhz.jet.enabled=true (for Hazelcast embedded, works only when loading config via Config.load)
  - Add environment variable: HZ_JET_ENABLED=true (recommended when running container image. For Hazelcast embedded, works only when loading config via Config.load)
12:08:12,737  INFO |testAuthenticationExceptionDoesNotLeakInfo| - [MetricsConfigHelper] main - [127.0.0.1]:5701 [45c8778d-641e-49aa-a0f7-5e075f9348dd] [5.3.0-SNAPSHOT] Collecting debug metrics and sending to diagnostics is enabled
12:08:12,742  INFO |testAuthenticationExceptionDoesNotLeakInfo| - [Node] main - [127.0.0.1]:5701 [45c8778d-641e-49aa-a0f7-5e075f9348dd] [5.3.0-SNAPSHOT] Using Multicast discovery
12:08:12,742  WARN |testAuthenticationExceptionDoesNotLeakInfo| - [CPSubsystem] main - [127.0.0.1]:5701 [45c8778d-641e-49aa-a0f7-5e075f9348dd] [5.3.0-SNAPSHOT] CP Subsystem is not enabled. CP data structures will operate in UNSAFE mode! Please note that UNSAFE mode will not provide strong consistency guarantees.
12:08:12,745  INFO |testAuthenticationExceptionDoesNotLeakInfo| - [Diagnostics] main - [127.0.0.1]:5701 [45c8778d-641e-49aa-a0f7-5e075f9348dd] [5.3.0-SNAPSHOT] Diagnostics disabled. To enable add -Dhazelcast.diagnostics.enabled=true to the JVM arguments.
12:08:12,745  INFO |testAuthenticationExceptionDoesNotLeakInfo| - [LifecycleService] main - [127.0.0.1]:5701 [45c8778d-641e-49aa-a0f7-5e075f9348dd] [5.3.0-SNAPSHOT] [127.0.0.1]:5701 is STARTING
12:08:14,925  INFO |testAuthenticationExceptionDoesNotLeakInfo| - [ClusterService] main - [127.0.0.1]:5701 [45c8778d-641e-49aa-a0f7-5e075f9348dd] [5.3.0-SNAPSHOT] 

Members {size:1, ver:1} [
	Member [127.0.0.1]:5701 - 8496403a-946e-4bf0-b701-6d02b4735d0c this
]

12:08:14,926  INFO |testAuthenticationExceptionDoesNotLeakInfo| - [LifecycleService] main - [127.0.0.1]:5701 [45c8778d-641e-49aa-a0f7-5e075f9348dd] [5.3.0-SNAPSHOT] [127.0.0.1]:5701 is STARTED
12:08:14,929 ERROR |testAuthenticationExceptionDoesNotLeakInfo| - [MapGetMessageTask] hz.gifted_leakey.partition-operation.thread-0 - [127.0.0.1]:5701 [45c8778d-641e-49aa-a0f7-5e075f9348dd] [5.3.0-SNAPSHOT] Client ClientEndpoint{connection=Connection[id=1, /127.0.0.1:5701->/127.0.0.1:37210, qualifier=null, endpoint=null, remoteUuid=null, alive=true, connectionType=NONE, planeIndex=-1], clientUuid=null, clientName=null, authenticated=false, clientVersion=Unknown, creationTime=1666267694929, latest clientAttributes=null, labels=null} must authenticate before any operation.
12:08:14,929  INFO |testAuthenticationExceptionDoesNotLeakInfo| - [TcpServerConnection] hz.gifted_leakey.partition-operation.thread-0 - [127.0.0.1]:5701 [45c8778d-641e-49aa-a0f7-5e075f9348dd] [5.3.0-SNAPSHOT] Connection[id=1, /127.0.0.1:5701->/127.0.0.1:37210, qualifier=null, endpoint=null, remoteUuid=null, alive=false, connectionType=NONE, planeIndex=-1] closed. Reason: Authentication failed. Client ClientEndpoint{connection=Connection[id=1, /127.0.0.1:5701->/127.0.0.1:37210, qualifier=null, endpoint=null, remoteUuid=null, alive=true, connectionType=NONE, planeIndex=-1], clientUuid=null, clientName=null, authenticated=false, clientVersion=Unknown, creationTime=1666267694929, latest clientAttributes=null, labels=null} must authenticate before any operation.
12:08:14,979  INFO |testAuthenticationExceptionDoesNotLeakInfo| - [LifecycleService] main - [127.0.0.1]:5701 [45c8778d-641e-49aa-a0f7-5e075f9348dd] [5.3.0-SNAPSHOT] [127.0.0.1]:5701 is SHUTTING_DOWN
12:08:14,979  INFO |testAuthenticationExceptionDoesNotLeakInfo| - [Node] main - [127.0.0.1]:5701 [45c8778d-641e-49aa-a0f7-5e075f9348dd] [5.3.0-SNAPSHOT] Shutting down multicast service...
12:08:14,980  INFO |testAuthenticationExceptionDoesNotLeakInfo| - [Node] main - [127.0.0.1]:5701 [45c8778d-641e-49aa-a0f7-5e075f9348dd] [5.3.0-SNAPSHOT] Shutting down connection manager...
12:08:14,981  INFO |testAuthenticationExceptionDoesNotLeakInfo| - [Node] main - [127.0.0.1]:5701 [45c8778d-641e-49aa-a0f7-5e075f9348dd] [5.3.0-SNAPSHOT] Shutting down node engine...
12:08:14,983  INFO |testAuthenticationExceptionDoesNotLeakInfo| - [NodeExtension] main - [127.0.0.1]:5701 [45c8778d-641e-49aa-a0f7-5e075f9348dd] [5.3.0-SNAPSHOT] Destroying node NodeExtension.
12:08:14,983  INFO |testAuthenticationExceptionDoesNotLeakInfo| - [Node] main - [127.0.0.1]:5701 [45c8778d-641e-49aa-a0f7-5e075f9348dd] [5.3.0-SNAPSHOT] Hazelcast Shutdown is completed in 4 ms.
12:08:14,983  INFO |testAuthenticationExceptionDoesNotLeakInfo| - [LifecycleService] main - [127.0.0.1]:5701 [45c8778d-641e-49aa-a0f7-5e075f9348dd] [5.3.0-SNAPSHOT] [127.0.0.1]:5701 is SHUTDOWN

```
</details>

### PR DESCRIPTION
**Pull request by srknzl**
The test failure was due to the client connection being closed before the exception message was able to be written. Upon closing, write queues of the connection being drained and this means no exception message will be sent to the client. To fix this, I used assertTrueEventually in the test.

See debug ss: 
Screenshot from 2022-11-17 09-05-48

Alternative ways of fixing this test failure is that:

1. We can delay closing the connection in AbstractMessageTask, (we think this is not the ideal solution)
2. We can change the logic so that we dont close the client connection at all. 

Will be backported to 5.2.z as well upon merging. 

Fixes #22564

Breaking changes (list specific methods/types/messages):
* None

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible
- [x] Send backports/forwardports if fix needs to be applied to past/future releases
- [x] New public APIs have `@Nonnull/@Nullable` annotations
- [x] New public APIs have `@since` tags in Javadoc

### CODE REVIEW

**Comment by srknzl** [comment]
@mdumandag Changed as you suggested.

File: hazelcast/src/test/java/com/hazelcast/client/AuthenticationInformationLeakTest.java [hunk 1]
```diff
@@ -167,11 +169,17 @@ private ClientMessage getKeyValue(SerializationService ss, OutputStream os, Inpu
...
-        ClientMessage res = ClientTestUtil.readResponse(is);
-        if (res.getMessageType() != ErrorsCodec.EXCEPTION_MESSAGE_TYPE) {
-            assertEquals(res.getMessageType(), MapGetCodec.RESPONSE_MESSAGE_TYPE);
+        AtomicReference<ClientMessage> res = new AtomicReference<>();
+        // The server sends an exception message and immediately closes the connection without waiting the message to
+        // be written. It's possible that we receive EOF from socket, which will make ClientTestUtil.readResponse throw.
+        // Therefore, we use assertTrueEventually here.
+        assertTrueEventually(() -> {
+            res.set(ClientTestUtil.readResponse(is));
+        });
```
Comment (mdumandag) [inline hazelcast/src/test/java/com/hazelcast/client/AuthenticationInformationLeakTest.java:178]: How is this going to work when the server closes the socket? You will not be able to read the message, and I assume `ClientTestUtil.readResponse` will throw an `EOFException`.

Comment (srknzl) [reply hazelcast/src/test/java/com/hazelcast/client/AuthenticationInformationLeakTest.java:178]: Right, we should put a assertTrueEventually to the general request-response logic.

Comment (srknzl) [reply hazelcast/src/test/java/com/hazelcast/client/AuthenticationInformationLeakTest.java]: Moved assertTrueEventually to the test logic in a way that it covers socket creation, doing the request, getting the response and asserting.

File: hazelcast/src/test/java/com/hazelcast/client/AuthenticationInformationLeakTest.java [hunk 2]
```diff
@@ -87,29 +88,34 @@ public void cleanUp() {
...
-            socket.setReuseAddress(true);
-            socket.connect(endpoint);
-            ClientMessage res;
-            try (OutputStream os = socket.getOutputStream(); InputStream is = socket.getInputStream()) {
-                os.write(CLIENT_BINARY.getBytes(StandardCharsets.UTF_8));
-                res = getKeyValue(ss, os, is);
+        // The server sends an exception message and immediately closes the connection without waiting the message to
+        // be written. It's possible that we receive EOF from socket, which will make ClientTestUtil.readResponse throw.
+        // Therefore, we use assertTrueEventually here.
+        assertTrueEventually(() -> {
```
Comment (mdumandag) [inline hazelcast/src/test/java/com/hazelcast/client/AuthenticationInformationLeakTest.java:97]: The assertTrueEventually still does not solve the problem we have. If the lambda throws any exception other than the assertion exception, it terminates by rethrowing that exception. Here, we will likely to get some sort of IOException from `res = getKeyValue(ss, os, is);` line. And, when we get that, the test will fail like before.

Comment (srknzl) [reply hazelcast/src/test/java/com/hazelcast/client/AuthenticationInformationLeakTest.java]: Makes sense. Did a refactor. can you check now?

**Review by mdumandag — APPROVED** [review]

**Pull request merged**